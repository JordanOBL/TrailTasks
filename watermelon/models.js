import {Model, Q} from '@nozbe/watermelondb';

import {
  relation,
  immutableRelation,
  children,
  field,
  text,
  writer,
  date,
  lazy,
} from '@nozbe/watermelondb/decorators';

import formatDateTime from '../helpers/formatTime';

export class Park extends Model {
  static table = 'parks';
  static associations = {
    trails: {type: 'has_many', foreignKey: 'park_id'},
    parks_states: {type: 'has_many', foreignKey: 'park_id'},
  };

  @field('park_name') parkName;
  @field('park_type') parkType;
  @field('park_image_url') parkImageUrl;

  @children('trails') trails;
}

export class Trail extends Model {
  static table = 'trails';
  static associations = {
    parks: {type: 'belongs_to', key: 'park_id'},
    users: {type: 'has_many', foreignKey: 'trail_id'},
    completed_hikes: {type: 'has_many', foreignKey: 'trail_id'},
    queued_trails: {type: 'has_many', foreignKey: 'trail_id'},
    basic_subscription_trails: {type: 'has_many', foreignKey: 'trail_id'},
  };
  //fields

  @field('trail_name') trailName;
  @field('trail_distance') trailDistance;
  @field('trail_lat') trailLat;
  @field('trail_long') trailLong;
  @field('trail_difficulty') trailDifficulty;
  @field('park_id') parkId;
  @field('trail_image_url') trailImageUrl;
  @field('all_trails_url') allTrailsUrl;
  @field('nps_url') npsUrl;
  @field('hiking_project_url') hikingProjectUrl;
  @field('trail_elevation') trailElevation;

  @relation('parks', 'park_id') park;

  ////?possibly show all uses currently hiking the trail at the time
  @children('users') users;
  @children('completed_hikes') completedHikes;
  @children('queued_trails') queuedTrails;
  @children('basic_subscription_trails') basicSubscriptionTrails;

  //Add To Hiking Queue
  @writer async addToQueuedTrails({userId}) {
    const addedHike = await this.collections
      .get('queued_trails')
      .create((hike) => {
        hike.trailId = this.id;
        hike.userId = userId;
      });
    return addedHike;
  }

  @writer async deleteFromQueuedTrails({userId}) {
    const deleteThisHike = await this.collections
      .get('queued_trails')
      .query(Q.and(Q.where('trail_id', this.id), Q.where('user_id', userId)));
    console.log('in queued Trails');
    await deleteThisHike[0].markAsDeleted();
  }
}

export class Basic_Subscription_Trail extends Model {
  static table = 'basic_subscription_trails';
  static associations = {
    trails: {type: 'belongs_to', key: 'trail_id'},
  };
  //fields
  @field('trail_id') trailId;
  @relation('trails', 'trail_id') trail;
  @children('trails') trails;
}

export class User extends Model {
  static table = 'users';
  static associations = {
    users_badges: {type: 'has_many', foreignKey: 'user_id'},
    users_achievements: {type: 'has_many', foreignKey: 'user_id'},
    users_sessions: {type: 'has_many', foreignKey: 'user_id'},
    trails: {type: 'belongs_to', key: 'trail_id'},
    completed_hikes: {type: 'has_many', foreignKey: 'user_id'},
    queued_trails: {type: 'has_many', foreignKey: 'user_id'},
    users_miles: {type: 'has_many', foreignKey: 'user_id'},
    users_subscriptions: {type: 'has_many', foreignKey: 'user_id'},
  };

  @field('username') username;
  @field('first_name') firstName;
  @field('last_name') lastName;
  @field('email') email;
  @field('password') password;
  @field('push_notifications_enabled') pushNotificationsEnabled;
  @field('theme_preference') themePreference;
  @field('trail_id') trailId;
  @field('trail_progress') trailProgress;
  @field('trail_started_at') trailStartedAt;
  @field('is_subscribed') isSubscribed;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @relation('trails', 'trail_id') trail;

  @children('users_sessions') usersSessions;
  @children('users_badges') usersBadges;
  @children('users_achievements') usersAchievements;
  @children('users_miles') usersMiles;
  @children('completed_hikes') completedHikes;
  @children('queued_trails') queuedTrails;
  @children('users_subscriptions') usersSubscriptions;

  @lazy userMiles = this.usersMiles.extend(Q.where('user_id', this.id));
  @lazy userSessionsWithCategory = this.usersSessions.extend(
    Q.on('session_categories', 'id', 'session_category_id')
  );

  // getUser
  @writer async getUser() {
    return {
      firstName: this.firstName,
      lastName: this.lastName,
      username: this.username,
      email: this.email,
      id: this.id,
      password: this.password,
      push_notifications_enabled: this.pushNotificationsEnabled,
      theme_preference: this.themePreference,
      trail_id: this.trailId,
      trail_progress: this.trailProgress,
      trail_started_at: this.trailStartedAt,
      updated_at: this.updatedAt,
      created_at: this.createdAt,
    };
  }

  @writer async increaseDistanceHikedWriter({user, userMiles, userSession}) {
    await this.batch(
      user.prepareUpdate((user) => {
        user.trailProgress = (Number(user.trailProgress) + 0.01).toFixed(2);
      }),
      userMiles.prepareUpdate((userMiles) => {
        userMiles.totalMiles = (Number(userMiles.totalMiles) + 0.01).toFixed(2);
      }),
      userSession.prepareUpdate((userSession) => {
        userSession.totalDistanceHiked = (
          Number(userSession.totalDistanceHiked) + 0.01
        ).toFixed(2);
      })
    );
  }

  //update User Trail
  //miles: number
  @writer async updateTrailProgress({miles}) {
    return await this.update(() => {
      this.trailProgress = (Number(this.trailProgress) + miles).toFixed(2);
    });
  }
  //update User Trail
  @writer async updateUserTrail({trailId, trailStartedAt}) {
    return await this.update(() => {
      this.trailId = trailId;
      this.trailProgress = '0.00';
      this.trailStartedAt = trailStartedAt;
    });
  }
  //Add User`
  @writer async addUser(
    username,
    firstName,
    lastName,
    email,
    password,
    trailStartedAt
  ) {
    const newUser = await this.collections.get('users').create((user) => {
      user.username = username;
      user.firstName = firstName;
      user.lastName = lastName;
      user.email = email;
      user.password = password;
      user.pushNotificationsEnabled = true;
      user.themePreference = 'light';
      user.trailId = '1';
      user.trailProgress = '0.0';
      user.traiStartedAt = trailStartedAt;
    });
    return newUser;
  }
  //add user miles
  @writer async addUserMile() {
    return await this.collections.get('users_miles').create((user_miles) => {
      user_miles.user_id.set(this);
      user_miles.totalMiles = '0.00';
    });
  }

  //add user achievnments Batch
  // @writer async unlockAchievements(completedAchievementIds) {
  //   const unlockedAchievements = await completedAchievementIds.map(
  //     (achievementId) =>
  //       this.collections
  //         .get('users_achievements')
  //         .prepareCreate((user_achievement) => {
  //           user_achievement.userId = this.userId;
  //           user_achievement.achievementId = achievementId;
  //         })
  //   );
  //   await this.database.batch(...unlockedAchievements);
  //   return unlockedAchievements;
  // }
  @writer async unlockAchievements(userId, completedAchievementIds) {
    try {
      console.debug('Unlocking achievements:', completedAchievementIds);
      const unlockedAchievements = await Promise.all(
        completedAchievementIds.map(async (achievementId) => {
          const newAchievement = this.collections
            .get('users_achievements')
            .prepareCreate((user_achievement) => {
              user_achievement.userId = userId;
              user_achievement.achievementId = achievementId;
              user_achievement.completedAt = formatDateTime(new Date());
            });
          return newAchievement;
        })
      );
      await this.database.batch(...unlockedAchievements);
      console.debug(
        'Successfully unlocked achievements:',
        unlockedAchievements
      );
      return unlockedAchievements;
    } catch (err) {
      console.error('Error in unlockAchievements:', err);
      throw err; // Rethrow the error to handle it at the higher level
    }
  }

  @writer async updateTotalUserMiles({miles}) {
    return await this.usersMiles.update((userMile) => {
      userMile.totalMiles = (Number(userMile.totalMiles) + miles).toFixed(2);
    });
  }
  //add User Session
  @writer async addUserSession({
    sessionName,
    sessionDescription,
    sessionCategoryId,
  }) {
    return await this.collections
      .get('users_sessions')
      .create((user_session) => {
        user_session.user_id.set(this);
        user_session.sessionName = sessionName;
        user_session.sessionDescription = sessionDescription;
        user_session.totalMilesHiked = '0.00';
        user_session.totalSessionTime = 0;
        user_session.sessionCategoryId = sessionCategoryId;
      });
  }
  //create completed_hike
  @writer async addCompletedHike({
    trailId,
    bestCompletedTime,
    firstCompletedAt,
    lastCompletedAt,
  }) {
    return await this.collections
      .get('completed_hikes')
      .create((completedHike) => {
        completedHike.userId = this.id;
        completedHike.trailId = trailId;
        completedHike.bestCompletedTime = bestCompletedTime;
        completedHike.firstCompletedAt = firstCompletedAt;
        completedHike.lastCompletedAt = lastCompletedAt;
      });
  }
  //update completed hike
  @writer async updateCompletedHike({
    completedHikeId,
    bestCompletedTime,
    lastCompletedAt,
  }) {
    const completedHike = await this.collections
      .get('completed_hikes')
      .find(completedHikeId);
    // eslint-disable-next-line no-shadow
    return await completedHike.update((completedHike) => {
      completedHike.bestCompletedTime = bestCompletedTime;
      completedHike.lastCompletedAt = lastCompletedAt;
    });
  }

  //create completed_hike
  @writer async hasTrailBeenCompleted() {
    const completedHike = await this.collections
      .get('completed_hikes')
      .query(
        Q.and(Q.where('user_id', this.id), Q.where('trail_id', this.trailId))
      );
    return completedHike[0];
  }
}

export class Park_State extends Model {
  static table = 'park_states';
  static associations = {
    parks: {type: 'belongs_to', key: 'park_id'},
  };

  @field('park_id') parkId;
  @field('state_code') stateCode;
  @field('state_name') stateName;

  @immutableRelation('parks', 'park_id') park;
}
export class Badge extends Model {
  static table = 'badges';
  static associations = {
    users_badges: {type: 'has_many', foreignKey: 'badge_id'},
  };

  @field('badge_name') badgeName;
  @field('badge_description') badgeDescription;
  @field('badge_image_url') badgeImageUrl;

  @immutableRelation('users_badges', 'badge_id') badge;

  @children('users_badges') users_badges;

  @lazy
  badgeEarners = this.collections
    .get('users')
    .query(Q.on('users_badges', 'badge_id', this.id));
}

export class Achievement extends Model {
  static table = 'achievements';
  static associations = {
    users_achievements: {type: 'has_many', foreignKey: 'achievement_id'},
  };
  @field('achievement_name') achievementName;
  @field('achievement_description') achievementDescription;
  @field('achievement_image_url') achievementImageUrl;
  @field('type') achievementType;
  @field('condition') achievementCondition;

  @immutableRelation('users_achievements', 'achievement_id') achievement;

  @children('users_achievements') usersAchievements;

  @lazy
  achievementEarners = this.collections
    .get('users')
    .query(Q.on('users_achievements', 'achievement_id', this.id));
}
export class User_Achievement extends Model {
  static table = 'users_achievements';
  static associations = {
    achievements: {type: 'belongs_to', key: 'achievement_id'},
    users: {type: 'belongs_to', key: 'user_id'},
  };
  @field('user_id') userId;
  @field('achievement_id') achievementId;
  @field('completed_at') completedAt;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @immutableRelation('users', 'user_id') user;
  @immutableRelation('achievements', 'achievement_id') achievement;

  @children('users') users;
  @children('achievements') achievements;
}

export class Completed_Hike extends Model {
  static table = 'completed_hikes';
  static associations = {
    users: {type: 'belongs_to', key: 'user_id'},
    trails: {type: 'belongs_to', key: 'trail_id'},
  };

  @field('user_id') userId;
  @field('trail_id') trailId;
  @field('first_completed_at') firstCompletedAt;
  @field('last_completed_at') lastCompletedAt;
  @field('best_completed_time') bestCompletedTime;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @immutableRelation('users', 'user_id') user;
  @immutableRelation('trails', 'trail_id') trail;

  @children('users') users;
  @children('trails') trails;

  // @lazy extendedTrail = this.trails.extend(Q.on('parks', 'park_id', 'park_id'));
}
export class Queued_Trail extends Model {
  static table = 'queued_trails';
  static associations = {
    trails: {type: 'belongs_to', key: 'trail_id'},
    users: {type: 'belongs_to', key: 'user_id'},
  };

  @field('user_id') userId;
  @field('trail_id') trailId;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @relation('users', 'user_id') user;
  @relation('trails', 'trail_id') trail;

  @children('users') users;
  @children('trails') trails;

  @writer async deleteTrailFromQueue() {
    await this.markAsDeleted();
  }
}
export class User_Miles extends Model {
  static table = 'users_miles';
  static associations = {
    users: {type: 'belongs_to', key: 'user_id'},
  };

  @field('user_id') userId;
  @field('total_miles') totalMiles;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @immutableRelation('users', 'user_id') user;

  @children('users') users;

  @writer async updateTotalMiles({miles}) {
    return await this.update(() => {
      this.totalMiles = (Number(this.totalMiles) + miles).toFixed(2);
    });
  }
}
export class User_Badge extends Model {
  static table = 'users_badges';
  static associations = {
    users: {type: 'belongs_to', key: 'user_id'},
    badges: {type: 'belongs_to', key: 'badge_id'},
  };
  @field('user_id') userId;
  @field('badge_id') badgeId;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @immutableRelation('users', 'user_id') user;
  @immutableRelation('badges', 'badge_id') badge;

  @children('users') users;
  @children('badges') badges;
}

export class Session_Category extends Model {
  static table = 'session_categories';
  static associations = {
    users_sessions: {type: 'has_many', foreignKey: 'session_category_id'},
  };

  @field('session_category_name') sessionCategoryName;

  @children('users_sessions') usersSessions;
}

export class User_Session extends Model {
  static table = 'users_sessions';
  static associations = {
    users: {type: 'belongs_to', key: 'user_id'},
    session_categories: {
      type: 'belongs_to',
      key: 'session_category_id',
    },
  };

  @field('user_id') userId;
  @text('session_name') sessionName;
  @text('session_description') sessionDescription;
  @field('session_category_id') sessionCategoryId;
  @field('date_added') dateAdded;
  @field('total_session_time') totalSessionTime;
  @field('total_distance_hiked') totalDistanceHiked;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @relation('users', 'user_id') user;
  @immutableRelation('session_categories', 'session_category_id')
  sessionCategory;

  @children('users') users;

  @writer async updateTotalDistanceHiked({miles}) {
    return await this.update(() => {
      this.totalDistanceHiked = (
        Number(this.totalDistanceHiked) + miles
      ).toFixed(2);
    });
  }

  @writer async updateTotalSessionTime() {
    return await this.update(() => {
      this.totalSessionTime = this.totalSessionTime + 1;
    });
  }
}

export class Subscription extends Model {
  static table = 'users_subscriptions';
  static associations = {
    users: {type: 'belongs_to', key: 'user_id'},
  };
  @field('user_id') userId;
  @field('is_active') isActive;
  @field('expires_at') expiresAt;
  @date('created_at') createdAt;
  @date('updated_at') updatedAt;

  @relation('users', 'user_id') user;

  @children('users') users;
}
