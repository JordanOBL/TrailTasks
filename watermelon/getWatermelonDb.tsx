// import * as React from 'react';
import {Database} from '@nozbe/watermelondb';
import SQLiteAdapter from '@nozbe/watermelondb/adapters/sqlite';
import schema from './schema';

// import your Models here
import {
  Park,
  Trail,
  User,
  Park_State,
  Badge,
  Achievement,
  User_Achievement,
  Completed_Hike,
  Hiking_Queue,
  User_Miles,
  User_Badge,
  Session_Category,
  User_Session,
} from './models';


const adapter = new SQLiteAdapter({
  schema,
  dbName: 'TrailTasks',
});

export const watermelonDatabase = new Database({
  adapter,
  modelClasses: [
    Park,
    Trail,
    User,
    Park_State,
    Badge,
    Achievement,
    User_Achievement,
    Completed_Hike,
    Hiking_Queue,
    User_Miles,
    User_Badge,
    Session_Category,
    User_Session,
  ],
});



export default watermelonDatabase;
