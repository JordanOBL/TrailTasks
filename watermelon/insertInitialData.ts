import {Database, Q} from '@nozbe/watermelondb';

async function insertInitialData(watermelonDatabase: Database) {
  try {
    const trailsinDB = watermelonDatabase
      .get('trails')
      .query(Q.where('trail_id', 5))
      .fetch();
    // const trails = await trailsinDB.query().fetch();
    return trailsinDB;
  } catch (error) {
    console.error('Error while inserting initial data:', error);
  }
}

export default insertInitialData;

// {"_cache": {"_debugCollection": [Circular], "map": Map {}, "recordInsantiator": [Function anonymous], "tableName": "trails"}, "_subscribers": [], "changes": {"closed": false, "currentObservers": null, "hasError": false, "isStopped": false, "observers": [], "thrownError": null}, "database": {"_isBeingReset": false, "_isBroken": false, "_pendingNotificationBatches": 0, "_pendingNotificationChanges": [], "_resetCount": 0, "_subscribers": [], "_workQueue": {"_db": [Circular], "_queue": [Array], "_subActionIncoming": false}, "adapter": {"underlyingAdapter": [SQLiteAdapter]}, "collections": {"map": [Object]}, "schema": {"tables": [Object], "unsafeSql": undefined, "version": 1}}, "modelClass": [Function Trail]}
