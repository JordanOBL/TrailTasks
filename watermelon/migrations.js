// app/model/migrations.js

import { schemaMigrations, createTable, addColumns } from '@nozbe/watermelondb/Schema/migrations'

export default schemaMigrations({
  migrations: [
    {
      // ⚠️ Set this to a number one larger than the current schema version
      toVersion: 2,
      steps: [
        createTable({
          name: 'users_friends',
          columns: [
            { name: 'user_id', type: 'string', isIndexed: true },
            { name: 'friend_id', type: 'string', isIndexed: true },
            { name: 'created_at', type: 'number' },
            { name: 'updated_at', type: 'number' },
          ],
        }), 
        createTable({
          name: 'cached_friends',
          columns: [
            { name: 'user_id', type: 'string', isIndexed: true },
            { name: 'friend_id', type: 'string', isIndexed: true },
            { name: 'total_miles', type: 'string', isOptional: false },
            { name: 'current_trail', type: 'string', isOptional: false }, // Friend status
            {name: 'username', type: 'string', isOptional: false},
            {name: 'room_id', type: 'string', isOptional: true},
            { name: 'created_at', type: 'number' },
            { name: 'updated_at', type: 'number' },
          ],
        }), 
        addColumns({
          table: 'users',
          columns: [
            {name: 'room_id', type: 'string', isOptional: true},
          ]
        })
      ],
    },
    {
      // ⚠️ Set this to a number one larger than the current schema version
      toVersion: 3,
      steps: [
        addColumns({
          table: 'cached_friends',
          columns: [
            { name: 'trail_progress', type: 'string', isOptional: false },
          ]
        })      
      ],
    }
  ],

});
