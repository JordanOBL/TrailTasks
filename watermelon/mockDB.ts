import {Database} from '@nozbe/watermelondb';
import LokiJSAdapter from '@nozbe/watermelondb/adapters/lokijs';

import mySchema from '../watermelon/schema';
import {
  Park,
  Trail,
  User,
  Park_State,
  Badge,
  Achievement,
  User_Achievement,
  Completed_Hike,
  Queued_Trail,
  User_Miles,
  User_Badge,
  Session_Category,
  User_Session,
  Subscription,
} from './models';

const adapter = new LokiJSAdapter({
  dbName: 'TestDb',
  schema: mySchema,
  useWebWorker: false,
  useIncrementalIndexedDB: true,
  extraLokiOptions: {
    autosave: false,
  },
});

export const testDb =  new Database({
  adapter,
  modelClasses: [
    Park,
    Trail,
    User,
    Park_State,
    Badge,
    Achievement,
    User_Achievement,
    Completed_Hike,
    Queued_Trail,
    User_Miles,
    User_Badge,
    Session_Category,
    User_Session,
    Subscription,
  ],
});
