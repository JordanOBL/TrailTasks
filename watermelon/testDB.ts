import {Database} from '@nozbe/watermelondb';
import SQLiteAdapter from '@nozbe/watermelondb/adapters/sqlite'

import mySchema from './schema';
import {
  Achievement,
  User_Completed_Trail,
  Park,
  Park_State,
  User_Queued_Trail,
  Session_Category,
  Subscription,
  Trail,
  User,
  User_Achievement,
  User_Purchased_Trail,
  User_Session,
  Addon,
  User_Addon,
  Session_Addon,
  User_Park
} from './models';

// Create the Node-based SQLite adapter
const adapter = new SQLiteAdapter({
  schema: mySchema,
  // Watermelon checks if we’re on RN or Node. 
  // Because we mocked index.native, it’ll use `better-sqlite3` in Node.
  synchronous: true, 
  jsi: false, // If your version of Watermelon complains about JSI, ensure it's off
})


export const testDb = new Database({
  adapter,
  modelClasses: [
  Achievement,
  User_Completed_Trail,
  Park,
  Park_State,
  User_Queued_Trail,
  Session_Category,
  Subscription,
  Trail,
  User,
  User_Achievement,
  User_Purchased_Trail,
  User_Session,
  Addon,
  User_Addon,
  Session_Addon,
  User_Park
  ],actionsEnabled: true,
});
