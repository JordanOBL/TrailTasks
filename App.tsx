import React, {useEffect} from "react";
//import type {PropsWithChildren} from 'react';
import RNFS from "react-native-fs";
import {
  SafeAreaView,
  Pressable,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
} from "react-native";
import {Colors} from "react-native/Libraries/NewAppScreen";
//import {Platform} from 'react-native';
//import useWatermelonDb from './watermelon/getWatermelonDb';
import watermelonDatabase from "./watermelon/getWatermelonDb";
import LoginScreen from "./Screens/LoginScreen";
import RegisterScreen from "./Screens/RegisterScreen";
import SyncIndicator from "./components/SyncIndicator";
import { sync } from './watermelon/sync';
import User from './watermelon/models'
import {hasUnsyncedChanges} from '@nozbe/watermelondb/sync';


// type SectionProps = PropsWithChildren<{
//   title: string;
// }>;

// function Section({children, title}: SectionProps): JSX.Element {
//   const isDarkMode = useColorScheme() === 'dark';
//   return (
//     <View style={styles.sectionContainer}>
//       <Text
//         style={[
//           styles.sectionTitle,
//           {
//             color: isDarkMode ? Colors.white : Colors.black,
//           },
//         ]}>
//         {title}
//       </Text>
//       <Text
//         style={[
//           styles.sectionDescription,
//           {
//             color: isDarkMode ? Colors.light : Colors.dark,
//           },
//         ]}>
//         {children}
//       </Text>
//     </View>
//   );
// }

function App(): JSX.Element {
  const [user, setUser] = React.useState<any>(null);
  const [isRegistering, setisRegistering] = React.useState<boolean>(true);
  //const watermelonDatabase = useWatermelonDb();
  const isDarkMode = useColorScheme() === "dark";

  const backgroundStyle = {
    backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
  };

  const checkForExistingUser = async () => {
    try {
      const userId = await watermelonDatabase.localStorage.get("user_id"); // string or undefined if no value for this key
      console.log("User info from local Storage", {userId});

      if (userId) {
        console.log("FOUND USER!");
        setUser({userId});
      }
      return;
    } catch (error) {
      console.error("Error in checkForUser function, app.tsx", error);
    }
  };

  // const getUserFirstName = async () => {
  //   const firstName = await watermelonDatabase
  //     .get('users')
  //     .query(Q.where('first_name', 'C'))
  //     .fetch();

  //   if (firstName.length > 0) {
  //     console.log(firstName);
  //   }
  // };
  const handleLogOut = async () => {
    try {
      await watermelonDatabase.localStorage.remove("user_id");
      await watermelonDatabase.localStorage.remove("username");
      setUser(null);
    } catch (error) {
      console.error("Error in handleLogOut function, app.tsx", error);
    }
  };
  const seedPgTables = async () => {
    try {
      console.log('seedingPgTables');
      const response = await fetch('http://localhost:5500/api/seed');
      const data = await response.json();
      console.log(data);
    } catch (error: any) {
      console.error(
        'Error in gettingPGTables function, app.tsx',
        error.message
      );
    }
  };
async function checkUnsyncedChanges() {
  const database = watermelonDatabase;
  await hasUnsyncedChanges({database});
}

  useEffect(() =>
  {
    
    // Do something with the Watermelon database instance
    const onLoad = async () => {
      try {
        const dbFilePath = `${RNFS.DocumentDirectoryPath}/TrailTasks.db`;
        console.log(`The database file is located at: ${dbFilePath}`);
         //await seedPgTables();
        console.log("Watermelon database:", watermelonDatabase);
        //console.log(watermelonDatabase.withChangesForTables(["parks"]));
        //await checkForExistingUser();
        await sync();
        // await checkUnsyncedChanges()
        // await getUserFirstName();
        // Find the location of the database file

        // Log the location of the database file to the console
        
      } catch (err: any) {
        console.log("Error in onload in APP useEffect", err.message);
      }
    };
    if (!user && watermelonDatabase) {
      onLoad();
    }
  }, [user]);

  return (
    <SafeAreaView style={[backgroundStyle, styles.container]}>
      <StatusBar
        barStyle={isDarkMode ? 'light-content' : 'dark-content'}
        backgroundColor={backgroundStyle.backgroundColor}
      />
      {/* <SyncIndicator /> */}
      {user && user.userId ? (
        <Text>LOGGED IN!</Text>
      ) : isRegistering ? (
        <RegisterScreen setUser={setUser} />
      ) : (
        <LoginScreen setUser={setUser} />
      )}
      <Pressable onPress={() => setisRegistering((prev: boolean) => !prev)}>
        <Text>{isRegistering ? 'Login' : 'Create an Account'}</Text>
      </Pressable>
      {user ? (
        <Pressable onPress={() => handleLogOut()}>
          <Text>Logout</Text>
        </Pressable>
      ) : (
        <></>
      )}
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: "600",
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: "400",
  },
  highlight: {
    fontWeight: "700",
  },
});

export default App;
