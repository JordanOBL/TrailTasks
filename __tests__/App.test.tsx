// __tests__/App.test.tsx
import React from 'react'
import renderer from 'react-test-renderer'
import App from '../App'
import { DatabaseProvider } from '@nozbe/watermelondb/react'
import { testDb } from '../watermelon/testDB'

// (Optional) Clean DB after each test
afterEach(async () => {
  await testDb.write(async () => {
    await testDb.unsafeResetDatabase()
  })
})

// A simple snapshot or integration test
it('renders correctly with real SQLite DB', async () => {
  // If the code inside <App /> calls WatermelonDB (including raw queries),
  // it will do so via the Node-based adapter now.
  const tree = renderer.create(
    <DatabaseProvider database={testDb}>
      <App />
    </DatabaseProvider>
  )

  // You might have an async operation that updates DB & re-renders,
  // so wrap in act if needed:
  // await renderer.act(async () => { ...some DB writes... })

  expect(tree.toJSON()).toMatchSnapshot()
})

