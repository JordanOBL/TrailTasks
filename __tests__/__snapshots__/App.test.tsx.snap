// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with real SQLite DB 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#F3F3F3",
      },
      {
        "backgroundColor": "rgb(31,33,35)",
        "flex": 1,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "black",
          "flex": 1,
          "padding": 16,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "alignItems": "center",
          "flexGrow": 1,
          "justifyContent": "center",
          "paddingVertical": 20,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "stretch",
              "justifyContent": "center",
              "marginBottom": 20,
              "width": "100%",
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/StopwatchLogo.png",
              }
            }
            style={
              {
                "borderRadius": 40,
                "height": 50,
                "width": 50,
              }
            }
          />
          <Text
            style={
              {
                "color": "white",
                "fontSize": 24,
                "fontWeight": "bold",
                "marginTop": 10,
              }
            }
          >
            New Hiker
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "stretch",
              "paddingHorizontal": 20,
              "width": "100%",
            }
          }
        >
          <View>
            <RCTSafeAreaView
              style={
                [
                  {
                    "backgroundColor": "black",
                    "flex": 1,
                  },
                  {
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 18,
                    "margin": 12,
                    "padding": 8,
                    "textAlign": "center",
                  }
                }
              >
                You neeed internet connection to register
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderRadius": 10,
                      "justifyContent": "center",
                      "marginVertical": 8,
                      "padding": 15,
                      "width": "100%",
                    },
                    {
                      "backgroundColor": "rgba(30, 139, 195, .7)",
                      "marginVertical": 12,
                    },
                  ]
                }
              >
                <Text
                  style={
                    {
                      "color": "black",
                      "fontSize": 20,
                      "fontWeight": "bold",
                    }
                  }
                >
                  Refresh Connection
                </Text>
              </View>
            </RCTSafeAreaView>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 10,
                    "justifyContent": "center",
                    "marginVertical": 8,
                    "padding": 15,
                    "width": "100%",
                  },
                  {
                    "backgroundColor": "rgb(7,254,213)",
                  },
                ]
              }
            >
              <Text
                style={
                  {
                    "color": "black",
                    "fontSize": 20,
                    "fontWeight": "bold",
                  }
                }
              >
                Login
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
